drivers/mcp9808: add temperature driver

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR add support for the [Microchip MCP9808](http://ww1.microchip.com/downloads/en/DeviceDoc/25095A.pdf) temperature sensor. The sensor uses I2C bus to communicate with a microcontroller.

The driver supports:
- basic polling
- wakeup/shutdown of the device for minimal power consumption
- saul interface

A test application is provided.

Interrupt mode could be added later but I wanted the driver to remain simple in a first PR.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

- Grab a MCP9808 and wire it to a board
- Build and flash the test application:
```
$ make BOARD=arduino-zero -C tests/driver_mcp9808 flash term
```
- Watch the temperature values being displayed every 2 seconds.

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

None

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

leandrolanzieri commented on drivers/mcp9808/mcp9808.c:
```diff
@@ -0,0 +1,168 @@
+
+    /* Reset config register to default values */
+    uint16_t init = 0;
+    if (_write16(dev, MCP9808_REG_CONFIG, &init) != MCP9808_OK) {
```
Here `_write16` is releasing the I2C bus and then it's being used again without acquiring. I think maybe the release in `_write16` could be removed and handled outside.

leandrolanzieri commented on drivers/mcp9808/mcp9808.c:
```diff
@@ -0,0 +1,168 @@
+
+    config &= ~MCP9808_REG_CONFIG_SHDN;
+    DEBUG("[mcp9808] wake_up: setting config to 0x%04X\n", config);
+    if (_write16(dev, MCP9808_REG_CONFIG, &config) != MCP9808_OK) {
```
`_write16` does not acquire the bus.

leandrolanzieri commented on drivers/mcp9808/mcp9808.c:
```diff
@@ -0,0 +1,168 @@
+
+    config |= MCP9808_REG_CONFIG_SHDN;
+    DEBUG("[mcp9808] shutdown: setting config to 0x%04X\n", config);
+    if (_write16(dev, MCP9808_REG_CONFIG, &config) != MCP9808_OK) {
```
Same as above.

leandrolanzieri commented on drivers/include/mcp9808.h:
```diff
@@ -0,0 +1,131 @@
+ * @brief   Read temperature value from the given MCP9808 device, returned in d°C
+ *
+ * @param[in] dev           Device descriptor of MCP9808 device
+ * @@param[out] temperature Temperature in d°C
```
```suggestion
 * @param[out] temperature Temperature in d°C
```

aabadie commented:
Thanks for reviewing @leandrolanzieri. I addressed your comments.

Here is the output of the test application:
```
$ make BOARD=arduino-zero -C tests/driver_mcp9808/ flash term
make: Entering directory '/home/aabadie/softs/src/riot/RIOT/tests/driver_mcp9808'
Building application "tests_driver_mcp9808" for "arduino-zero" with MCU "samd21".

"make" -C /home/aabadie/softs/src/riot/RIOT/boards/arduino-zero
"make" -C /home/aabadie/softs/src/riot/RIOT/core
"make" -C /home/aabadie/softs/src/riot/RIOT/cpu/samd21
"make" -C /home/aabadie/softs/src/riot/RIOT/cpu/cortexm_common
"make" -C /home/aabadie/softs/src/riot/RIOT/cpu/cortexm_common/periph
"make" -C /home/aabadie/softs/src/riot/RIOT/cpu/sam0_common
"make" -C /home/aabadie/softs/src/riot/RIOT/cpu/sam0_common/periph
"make" -C /home/aabadie/softs/src/riot/RIOT/cpu/samd21/periph
"make" -C /home/aabadie/softs/src/riot/RIOT/drivers
"make" -C /home/aabadie/softs/src/riot/RIOT/drivers/mcp9808
"make" -C /home/aabadie/softs/src/riot/RIOT/drivers/periph_common
"make" -C /home/aabadie/softs/src/riot/RIOT/sys
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/auto_init
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/div
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/isrpipe
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/newlib_syscalls_default
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/pm_layered
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/stdio_uart
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/tsrb
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/xtimer
   text	   data	    bss	    dec	    hex	filename
  12032	    140	   2664	  14836	   39f4	/home/aabadie/softs/src/riot/RIOT/tests/driver_mcp9808/bin/arduino-zero/tests_driver_mcp9808.elf
/home/aabadie/softs/src/riot/RIOT/dist/tools/edbg/edbg  -t atmel_cm0p -b -v -p -f /home/aabadie/softs/src/riot/RIOT/tests/driver_mcp9808/bin/arduino-zero/tests_driver_mcp9808.bin
Debugger: ATMEL EDBG CMSIS-DAP 00000000AZE000000166 02.01.0157 (S)
Clock frequency: 16.0 MHz
Target: SAM D21G18A (Rev D)
Programming................................................... done.
Verification................................................... done.
/home/aabadie/softs/src/riot/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200"
2018-12-19 14:42:07,870 - INFO # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
2018-12-19 14:42:10,374 - INFO # main(): This is RIOT! (Version: 2018.10-RC1-632-g3573db-pr/drivers/mcp9808)
2018-12-19 14:42:10,376 - INFO # MCP9808 test application
2018-12-19 14:42:10,377 - INFO # 
2018-12-19 14:42:10,379 - INFO # Initialization successful
2018-12-19 14:42:10,379 - INFO # 
2018-12-19 14:42:10,632 - INFO # Temperature [d°C]: 244
2018-12-19 14:42:12,890 - INFO # Temperature [d°C]: 245
2018-12-19 14:42:15,148 - INFO # Temperature [d°C]: 246
2018-12-19 14:42:17,405 - INFO # Temperature [d°C]: 288
2018-12-19 14:42:19,662 - INFO # Temperature [d°C]: 266
2018-12-19 14:42:21,920 - INFO # Temperature [d°C]: 258
2018-12-19 14:42:24,177 - INFO # Temperature [d°C]: 255
2018-12-19 14:42:25,588 - INFO # Exiting Pyterm
```
 
And using SAUL:
```
USEMODULE=mcp9808 make BOARD=arduino-zero -C examples/saul flash term
make: Entering directory '/home/aabadie/softs/src/riot/RIOT/examples/saul'
Building application "saul_example" for "arduino-zero" with MCU "samd21".

"make" -C /home/aabadie/softs/src/riot/RIOT/boards/arduino-zero
"make" -C /home/aabadie/softs/src/riot/RIOT/core
"make" -C /home/aabadie/softs/src/riot/RIOT/cpu/samd21
"make" -C /home/aabadie/softs/src/riot/RIOT/cpu/cortexm_common
"make" -C /home/aabadie/softs/src/riot/RIOT/cpu/cortexm_common/periph
"make" -C /home/aabadie/softs/src/riot/RIOT/cpu/sam0_common
"make" -C /home/aabadie/softs/src/riot/RIOT/cpu/sam0_common/periph
"make" -C /home/aabadie/softs/src/riot/RIOT/cpu/samd21/periph
"make" -C /home/aabadie/softs/src/riot/RIOT/drivers
"make" -C /home/aabadie/softs/src/riot/RIOT/drivers/mcp9808
"make" -C /home/aabadie/softs/src/riot/RIOT/drivers/periph_common
"make" -C /home/aabadie/softs/src/riot/RIOT/drivers/saul
"make" -C /home/aabadie/softs/src/riot/RIOT/sys
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/auto_init
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/auto_init/saul
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/fmt
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/isrpipe
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/newlib_syscalls_default
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/phydat
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/pm_layered
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/ps
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/saul_reg
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/shell
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/shell/commands
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/stdio_uart
"make" -C /home/aabadie/softs/src/riot/RIOT/sys/tsrb
   text	   data	    bss	    dec	    hex	filename
  16320	    504	   2660	  19484	   4c1c	/home/aabadie/softs/src/riot/RIOT/examples/saul/bin/arduino-zero/saul_example.elf
/home/aabadie/softs/src/riot/RIOT/dist/tools/edbg/edbg  -t atmel_cm0p -b -v -p -f /home/aabadie/softs/src/riot/RIOT/examples/saul/bin/arduino-zero/saul_example.bin
Debugger: ATMEL EDBG CMSIS-DAP 00000000AZE000000166 02.01.0157 (S)
Clock frequency: 16.0 MHz
Target: SAM D21G18A (Rev D)
Programming..................................................................... done.
Verification..................................................................... done.
/home/aabadie/softs/src/riot/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200"
2018-12-19 14:42:55,478 - INFO # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
2018-12-19 14:43:00,574 - INFO # main(): This is RIOT! (Version: 2018.10-RC1-632-g3573db-pr/drivers/mcp9808)
2018-12-19 14:43:00,576 - INFO # Welcome to RIOT!
2018-12-19 14:43:00,576 - INFO # 
2018-12-19 14:43:00,581 - INFO # Type `help` for help, type `saul` to see all SAUL devices
2018-12-19 14:43:00,581 - INFO # 
> saul
2018-12-19 14:43:02,413 - INFO #  saul
2018-12-19 14:43:02,415 - INFO # ID	Class		Name
2018-12-19 14:43:02,417 - INFO # #0	ACT_SWITCH	LED(orange)
2018-12-19 14:43:02,419 - INFO # #1	SENSE_TEMP	mcp9808
> saul read 1
2018-12-19 14:43:07,967 - INFO #  saul read 1
2018-12-19 14:43:07,971 - INFO # Reading from #1 (mcp9808|SENSE_TEMP)
2018-12-19 14:43:07,972 - INFO # Data:	     25.0°C
> 
2018-12-19 14:43:14,151 - INFO #  saul read 1
2018-12-19 14:43:14,154 - INFO # Reading from #1 (mcp9808|SENSE_TEMP)
2018-12-19 14:43:14,156 - INFO # Data:	     26.5°C
> 2018-12-19 14:43:16,963 - INFO # Exiting Pyterm
```

bergzand commented:
@aabadie Not 100% sure, but if I remember correctly, the mcp9808 is one of the devices compatible with the jc42 (generic) temperature sensor driver.

aabadie commented:
>  Not 100% sure, but if I remember correctly, the mcp9808 is one of the devices compatible with the jc42 (generic) temperature sensor driver.

Good point, I missed that. But the [doxygen documentation](https://riot-os.org/api/group__drivers__jc42.html) mentions MCP9804. I'll check

aabadie commented:
Ok, you are right, the sensor is part of the list: https://www.kernel.org/doc/Documentation/hwmon/jc42

This list should be more explicit in jc42 and the documentation improved. A lot of features are missing and some registers (temperature resolution) are not part of the standard.

I'll see if I can rework jc42 to better integrate that.

bergzand commented:
> This list should be more explicit in jc42 and the documentation improved. A lot of features are missing and some registers (temperature resolution) are not part of the standard.

Agreed,


> I'll see if I can rework jc42 to better integrate that.

That would be awesome!

leandrolanzieri commented:
@aabadie shall we close this in favor of a rework of the jc42 driver?

aabadie commented:
> shall we close this in favor of a rework of the jc42 driver?

I prefer to keep it open until I propose an update of the jc42 driver. This lies in my opened PRs and I kind of see it as a todo-list.
